### PR TITLE
[WIP] Improve building

### PR DESCRIPTION
**Pull request by tarsius**
@vermiculus "Rename Makefile to .travis.mk" breaks the build. Am I doing something wrong or does `emake.mk` not support this?

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ matrix:
     - env: EMACS_VERSION=snapshot
 
 before_install:
-  - wget "https://raw.githubusercontent.com/vermiculus/emake.el/${EMAKE_SHA1}/emake.mk"
+  - make emake.mk
```
Comment (vermiculus) [inline .travis.yml]: If you include an emake.mk target in the same makefile that includes emake.mk, that will always create an error the first time the makefile is run.

**Comment by vermiculus** [comment]
I can't identify locally any reason it shouldn't. I would assume that change would break cases where I'm assuming the calling pattern of Make, but I don't appear to be doing that anywhere.

I'll dig in a little more; are you sure it's that commit?

**Comment by vermiculus** [comment]
It would appear [the limitation may exist in emacs-travis.mk](https://travis-ci.org/magit/libegit2/jobs/450570516#L492-L495) rather than emake. L492 is emake deferring to emacs-travis to install emacs; the error happens within that Make.

**Comment by vermiculus** [comment]
If you sed-replace every `@make` in `.emake/emacs-travis.mk` with `@$(MAKE)`, what does that give you? (I'm hoping it'll give Make some memory of the `.travis.mk` that exists in the directory above.)

**Comment by vermiculus** [comment]
If I might ask a different question – what is driving this change?

**Comment by vermiculus** [comment]
I will think on this. There should be a way to do what you want (i.e., rename the makefile), but I'll have to dig into exactly why emacs-travis.mk is tripping on itself. I won't be able to get to that until tomorrow night, though -- this will require me to create a dummy fork for testing and I need to make sure there are no changes on my real fork :-)

That said, I didn't imagine emake being a separate build utility altogether -- I personally use it to build and test my packages locally since it removes duplication. Granted, there are other ways of handling that, and I'll cede that is a project management decision emake should not be assuming.

> `-import emake.mk`

Ooo, I'll have to add this to the documentation :-)

**Comment by tarsius** [comment]
> I will think on this. There should be a way to do what you want (i.e., rename the makefile), but I'll have to dig into exactly why emacs-travis.mk is tripping on itself.

My guess would be that a child `make` process inherits the value set using the `-f` argument. Maybe this is done by exporting a variable and changing that back would fix it. Or maybe you can override it by starting the child using `make -f Makefile [-C subdir] target`.

> That said, I didn't imagine emake being a separate build utility altogether -- I personally use it to build and test my packages locally since it removes duplication.

I am not opposed to that but given the project description, I didn't know whether that's a supported use case. And given that this particular Makefile simply did not work without Travis I decided to with what I assumed would be the easy option: moving it out of the way.

Anyway, I am actually considering to use `emake.mk` for testing Magit on Travis. Now that more dependencies are being added (some of which consist of more than one library) fetching the libraries by hand no longer is a reasonable approach.

**Comment by vermiculus** [comment]
> Would you like to implement that?

I've been wanting to do it for some time, but it felt presumptuous for such an established project. Given the invitation, though... expect a PR within the week :)

**Comment by TheBB** [comment]
I tried renaming the makefile in another project, and the same thing happened. I'm pretty sure the problem is here:

https://github.com/vermiculus/emake.el/blob/master/emake.mk#L112

This `make install-emacs` won't work if the file that includes `emake.mk` is named something other than `Makefile`.

@vermiculus Would it be possible to get this to invoke the correct makefile using something like the `MAKEFILE_LIST` variable? https://www.gnu.org/software/make/manual/html_node/Special-Variables.html

**Comment by vermiculus** [comment]
We're not using make directly there though, we're using `$(MAKE)`.

**Comment by vermiculus** [comment]
It's still pretty early here so I'm sorry if I'm just being bone-headed, but from what I see in the Travis log, we do get into the `install-emacs` target and the `make` call to `emacs-travis`

**Comment by tarsius** [comment]
@vermiculus Since you are working on Emake anyway... a friendly ping.

**Comment by vermiculus** [comment]
Thanks for bringing this back to my attention. Do you think vermiculus/emake.el#30 would resolve this concern?

> That way we can use "make" with a Makefile that is more suitable when building locally.

**Comment by vermiculus** [comment]
I don't have push access to this branch, but try this:

```diff
diff --git a/.travis.mk b/.travis.mk
index b9dd0b4..7043147 100644
--- a/.travis.mk
+++ b/.travis.mk
@@ -1,4 +1,4 @@
-EMAKE_SHA1       ?= 1b23379eb5a9f82d3e2d227d0f217864e40f23e0
+EMAKE_SHA1       ?= local-dev
 PACKAGE_BASENAME := libgit
```

Nor apparently do I know how to just copy a diff

**Comment by vermiculus** [comment]
I was able to give this a quick test and, while using a more recent version of emake works (specifically, the version that starts using EVM by default – but remember to also update `.travis.yml`!), it exposes an error in the CMake build:

```
mkdir -p build
cd build && cmake .. -DCMAKE_BUILD_TYPE=Debug && make
-- The C compiler identification is GNU 4.8.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- broken
CMake Error at /usr/local/cmake-3.9.2/share/cmake-3.9/Modules/CMakeTestCCompiler.cmake:51 (message):
  The C compiler "/usr/bin/cc" is not able to compile a simple test program.

  It fails with the following output:

   Change Dir: /home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeTmp

  

  Run Build Command:"/usr/bin/make" "cmTC_5b251/fast"

  make[1]: Entering directory
  `/home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeTmp'

  make -f .travis.mk -f CMakeFiles/cmTC_5b251.dir/build.make
  CMakeFiles/cmTC_5b251.dir/build

  make[2]: Entering directory
  `/home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeTmp'

  make[2]: .travis.mk: No such file or directory

  make[2]: *** No rule to make target `.travis.mk'.  Stop.

  make[2]: Leaving directory
  `/home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeTmp'

  make[1]: *** [cmTC_5b251/fast] Error 2

  make[1]: Leaving directory
  `/home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeTmp'

  

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:2 (enable_language)


-- Configuring incomplete, errors occurred!
See also "/home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeOutput.log".
See also "/home/travis/build/vermiculus/libegit2/build/CMakeFiles/CMakeError.log".
make: *** [build/libegit2.so] Error 1
```

https://travis-ci.org/vermiculus/libegit2/builds/497541106#L114

I don't have much experience with cmake, but I'm not sure why it would be searching for .travis.mk other than possibly being passed under `-f` in `$MAKE`.

**Comment by vermiculus** [comment]
It's been a long struggle, but I was able to get it working.

I re-implemented a very basic version of `emacs-travis.mk` that does not have the apparent limitation and have added it to the most recent version of EMake as a one-liner.

See [this working build](https://travis-ci.org/vermiculus/libegit2/builds/497952852?utm_source=github_status&utm_medium=notification).

**Pull request closed without merging**